native/cocos/renderer/pipeline/shadow/CSMLayers.h

```diff
@@ -0,0 +1,164 @@
+ http://www.cocos.com
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "../base/TypeDef.h"
+#include "../math/Mat4.h"
+#include "../../../core/geometry/Frustum.h"
+#include "pipeline/Define.h"
+#include "scene/Camera.h"
+#include "scene/DirectionalLight.h"
+#include "scene/Shadow.h"
+
+namespace cc {
+namespace pipeline {
+class PipelineSceneData;
+
+class ShadowTransformInfo {
+public:
+    explicit ShadowTransformInfo(uint level);
+    ~ShadowTransformInfo() = default;
+
+    inline uint getLevel() const { return _level; }
+
+    inline RenderObjectList &getShadowObjects() { return _shadowObjects; }
+    inline void setShadowObjects(RenderObjectList &&ro) { _shadowObjects = std::forward<RenderObjectList>(ro); }
+    inline void addShadowObject(RenderObject &&obj) { _shadowObjects.emplace_back(obj); }
+    inline void clearShadowObjects() { _shadowObjects.clear(); }
+
+    inline float getShadowCameraFar() const { return _shadowCameraFar; }
+    inline void setShadowCameraFar(float shadowCameraFar) { _shadowCameraFar = shadowCameraFar; }
+
+    inline const Mat4 &getMatShadowView() const { return _matShadowView; }
+    inline void setMatShadowView(const Mat4 &matShadowView) { _matShadowView = matShadowView; }
+
+    inline const Mat4 &getMatShadowProj() const { return _matShadowProj; }
+    inline void setMatShadowProj(const Mat4 &matShadowProj) { _matShadowProj = matShadowProj; }
+
+    inline const Mat4 &getMatShadowViewProj() const { return _matShadowViewProj; }
+    inline void setMatShadowViewProj(const Mat4 &matShadowViewProj) { _matShadowViewProj = matShadowViewProj; }
+
+    inline geometry::Frustum getValidFrustum() const { return _validFrustum; }
+    inline geometry::Frustum *getValidFrustumPtr() { return &_validFrustum; }
```

Comment:
Why need to return pointer as there is another function to return geometry::Frustum?